### PR TITLE
[Test] Use Ray client in XGBoost train_small release test

### PR DESCRIPTION
**Pull request by mwtian**
## Why are these changes needed?

This is an example of using Anyscale connect for a Ray release test. The script is tested against latest release of Ray, Anyscale CLI, and releaser from head. The command to run is the same as running other release tests:
```
$ python releaser/e2e.py --test-config ray/release/xgboost_tests/xgboost_tests.yaml --test-name train_small
```

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: release/xgboost_tests/app_config.yaml
Comment (AmeerHajAli) [inline release/xgboost_tests/app_config.yaml:1]: you can also use :1.4.0 if you want.

File: release/xgboost_tests/workloads/train_small.py
Comment (AmeerHajAli) [inline release/xgboost_tests/workloads/train_small.py:36]: alternatively, you can use the anyscale SDK to create a cluster env (look at the test_anyscale_development with connect) and reuse this cluster env in all the tests. This will significantly reduce the time it takes to run some of these tests.

**Review by ijrsvt — COMMENTED** [review]
This overall looks pretty good. Let's sort out 4931 soon!

File: release/xgboost_tests/workloads/train_small.py
Comment (krfricke) [reply release/xgboost_tests/workloads/train_small.py:36]: +1, we have the functionality in the e2e script already, and it will solve a couple of problems, e.g. rendering the yaml jinja2 template and also creating the test data (which is created in the docker image).

The name to the cluster env could then be passed via a command line argument

**Comment by krfricke** [comment]
Re: 

```
xgboost_tests $ python create_test_data.py ./data/classification.parquet  # no sure if this is right
```

The problem here is that this command needs to be run on all nodes (head node and worker nodes). This is because all nodes should have access to the same data, and for testing purposes it is easier to just store them on a local disk rather than on a NFS or S3.

Previously we ran this command in the startup commands, so the dataset was created and written on all nodes after they booted. We moved this to the app config in the current iteration, which makes much more sense as the data will be stored on a separate layer in the docker container. Some other tests are using larger test data sizes, and this saves a lot of time. 

IMO here we should also aim to include this in the app config. I'd strongly vote to keep creating the app config with the e2e script and then just passing the identifier to `connect()` (if this is a valid API call).

File: release/xgboost_tests/workloads/train_small.py
Comment (mwtian) [reply release/xgboost_tests/workloads/train_small.py]: Reusing the cluster created in the e2e script now.

File: release/xgboost_tests/workloads/train_small.py
Comment (mwtian) [reply release/xgboost_tests/workloads/train_small.py]: This logic is deleted now.

File: release/xgboost_tests/app_config.yaml
Comment (mwtian) [reply release/xgboost_tests/app_config.yaml]: Sounds good. I think if there is no other blocker, the nightly image may include newer versions of non-Ray dependencies, which could be more suitable for nightly / release tests.

**Review by AmeerHajAli — APPROVED** [review]

**Comment by mwtian** [comment]
> Re:
> 
> ```
> xgboost_tests $ python create_test_data.py ./data/classification.parquet  # no sure if this is right
> ```
> 
> The problem here is that this command needs to be run on all nodes (head node and worker nodes). This is because all nodes should have access to the same data, and for testing purposes it is easier to just store them on a local disk rather than on a NFS or S3.
> 
> Previously we ran this command in the startup commands, so the dataset was created and written on all nodes after they booted. We moved this to the app config in the current iteration, which makes much more sense as the data will be stored on a separate layer in the docker container. Some other tests are using larger test data sizes, and this saves a lot of time.
> 
> IMO here we should also aim to include this in the app config. I'd strongly vote to keep creating the app config with the e2e script and then just passing the identifier to `connect()` (if this is a valid API call).



> Re:
> 
> ```
> xgboost_tests $ python create_test_data.py ./data/classification.parquet  # no sure if this is right
> ```
> 
> The problem here is that this command needs to be run on all nodes (head node and worker nodes). This is because all nodes should have access to the same data, and for testing purposes it is easier to just store them on a local disk rather than on a NFS or S3.
> 
> Previously we ran this command in the startup commands, so the dataset was created and written on all nodes after they booted. We moved this to the app config in the current iteration, which makes much more sense as the data will be stored on a separate layer in the docker container. Some other tests are using larger test data sizes, and this saves a lot of time.
> 
> IMO here we should also aim to include this in the app config. I'd strongly vote to keep creating the app config with the e2e script and then just passing the identifier to `connect()` (if this is a valid API call).

Thanks Kai for the explanations. With project/releaser#11, generating the data would be done in app configs and this command is no longer necessary.

**Comment by krfricke** [comment]
Thanks - I'll take a look!

Please do not merge yet though, I want to make sure everything works with the current infrastructure. I'm happy to merge once it passes.

**Review by AmeerHajAli — APPROVED** [review]

**Comment by mwtian** [comment]
> Thanks - I'll take a look!
> 
> Please do not merge yet though, I want to make sure everything works with the current infrastructure. I'm happy to merge once it passes.

I see. I was thinking of merging the releaser change first. Let me make some changes to make sure this works with the current releaser.

File: release/xgboost_tests/xgboost_tests.yaml
Comment (krfricke) [inline release/xgboost_tests/xgboost_tests.yaml:14]: Doesn't this break execution as we don't have a shebang?

File: release/xgboost_tests/workloads/train_small.py
Comment (krfricke) [inline release/xgboost_tests/workloads/train_small.py:19]: `os.environ.get("RAY_JOB_NAME", "no_name")`?

**Review by krfricke — COMMENTED** [review]
just two questions

File: release/xgboost_tests/workloads/train_small.py
Comment (mwtian) [reply release/xgboost_tests/workloads/train_small.py]: Only set `job_name()` in client mode now, and default to `train_small`.

File: release/xgboost_tests/xgboost_tests.yaml
Comment (mwtian) [reply release/xgboost_tests/xgboost_tests.yaml]: I was thinking of handling `script` different when run as client, by passing `script` as one arg among the list of `args` to `subprocess.Run()`. Reverted now.

**Comment by mwtian** [comment]
The remaining requirement to make this test work in client mode is to use an Anyscale CLI that includes https://github.com/anyscale/product/commit/25a367f10f8f7d2863cdff1eb0e9ae1fa0bc2a81.

**Pull request merged**
